### DISCUSSION TITLE
Help with ELKO SmartPir med dimmer Plus PH - EKO06984 NHMOTION/DIMMER/1 Schneider

### DISCUSSION THREAD

**Discussion by prinsmikal** [Koenkk/zigbee2mqtt#15751] [Devices]
Elko is a norwegian brand. I found it uses hardware from Schneider. Found the same product with switch not dimmer in the schneider_electric.js  file. Tried to see if i could find some  more information in the elko.js 

I did some editing in the setup for 545D6306, in the schneider_electric.js file. 
Some of the functions work, but not the dimmer part, and not the PIR. I tried to get some data from other products that has PIR both from the schneider_electric.js and elko.js file but with no luck.

**This is the code I got working so far, I'll put the original code in the end.**

Code that does not kick errors at the moment:
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['NHMOTION/DIMMER/1'],
    model: 'EKO06984',
    vendor: 'Schneider Electric',
    description: 'SmartPir with dimmer Plus PH',
    fromZigbee: [...extend.light_onoff_brightness().fromZigbee, fz.wiser_lighting_ballast_configuration, fz.occupancy],
    toZigbee: [...extend.light_onoff_brightness().toZigbee, tz.ballast_config, tz.wiser_dimmer_mode],
    exposes: [e.occupancy(), e.illuminance(), e.light_brightness().withLevelConfig(),
        exposes.numeric('ballast_minimum_level', ea.ALL).withValueMin(1).withValueMax(254)
            .withDescription('Specifies the minimum light output of the ballast'),
        exposes.numeric('ballast_maximum_level', ea.ALL).withValueMin(1).withValueMax(254)
            .withDescription('Specifies the maximum light output of the ballast'),
        exposes.enum('dimmer_mode', ea.ALL, ['auto', 'rc', 'rl', 'rl_led'])
            .withDescription('Sets dimming mode to autodetect or fixed RC/RL/RL_LED mode (max load is reduced in RL_LED)')],
    whiteLabel: [{vendor: 'Elko', model: 'EKO07090'}],
    configure: async (device, coordinatorEndpoint, logger) => {
        await extend.light_onoff_brightness().configure(device, coordinatorEndpoint, logger);
        const endpoint = device.getEndpoint(3);
        await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff', 'genLevelCtrl', 'lightingBallastCfg']);
        await reporting.onOff(endpoint);
        await reporting.brightness(endpoint);
    },
    endpoint: (device) => {
        return {l1: 10, l2: 11};
    },
    meta: { multiEndpoint: true },
    configure: async(device, coordinatorEndpoint, logger) => {
        const endpoint1 = device.getEndpoint(1);
        const binds1 = ['genBasic', 'genIdentify', 'genOnOff'];
        await reporting.bind(endpoint1, coordinatorEndpoint, binds1);
        await reporting.onOff(endpoint1);
        // read switch state
        await endpoint1.read('genOnOff', ['onOff']);

        const endpoint37 = device.getEndpoint(37);
        const binds37 = ['msIlluminanceMeasurement', 'msOccupancySensing'];
        await reporting.bind(endpoint37, coordinatorEndpoint, binds37);
        await reporting.occupancy(endpoint37);
        await reporting.illuminance(endpoint37);
        // read occupancy_timeout
        await endpoint37.read('msOccupancySensing', ['pirOToUDelay']);
    },
};

module.exports = definition;

**Code I found in schneider_electric.js (this does not have a dimmer, but a switch. Some of the functions work, but not PIR, and Dimmer).**

    {
        zigbeeModel: ['NHMOTION/SWITCH/1'],
        model: '545D6306',
        vendor: 'Schneider Electric',
        description: 'LK FUGA Wiser wireless PIR with relay',
        fromZigbee: [fz.on_off, fz.illuminance, fz.occupancy, fz.occupancy_timeout],
        exposes: [e.switch().withEndpoint('l1'), e.occupancy(), e.illuminance_lux(), e.illuminance(),
            exposes.numeric('occupancy_timeout', ea.ALL).withUnit('second').withValueMin(0).withValueMax(3600)
                .withDescription('Time in seconds after which occupancy is cleared after detecting it')],
        toZigbee: [tz.on_off, tz.occupancy_timeout],
        endpoint: (device) => {
            return {'default': 37, 'l1': 1, 'l2': 37};
        },
        meta: {multiEndpoint: true},
        configure: async (device, coordinatorEndpoint, logger) => {
            const endpoint1 = device.getEndpoint(1);
            const binds1 = ['genBasic', 'genIdentify', 'genOnOff'];
            await reporting.bind(endpoint1, coordinatorEndpoint, binds1);
            await reporting.onOff(endpoint1);
            // read switch state
            await endpoint1.read('genOnOff', ['onOff']);

            const endpoint37 = device.getEndpoint(37);
            const binds37 = ['msIlluminanceMeasurement', 'msOccupancySensing'];
            await reporting.bind(endpoint37, coordinatorEndpoint, binds37);
            await reporting.occupancy(endpoint37);
            await reporting.illuminance(endpoint37);
            // read occupancy_timeout
            await endpoint37.read('msOccupancySensing', ['pirOToUDelay']);
        },
    },

**Comment by FSund** [comment]
I have two of the ELKO Smart motion dimmers (EKO06984), and found the following config to work okay (found it somewhere on a Facebook group). Most things are working (dimming etc), but I'm still trying to figure out how to get out the signal from the PIR sensor.

```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['NHMOTION/UNIDIM/1'],
    model: 'EKO06984',
    vendor: 'Elko',
    description: 'Elko SmartPir med dimmer Plus',
    fromZigbee: [fz.on_off, fz.brightness, fz.level_config, fz.schneider_lighting_ballast_configuration, fz.command_recall,
        fz.command_on, fz.command_off, fz.command_move, fz.command_stop],
    toZigbee: [tz.light_onoff_brightness, tz.level_config, tz.ballast_config, tz.schneider_dimmer_mode],
    endpoint: (device) => {
        return {'l1': 3, 's1': 21, 's2': 22, 's3': 23, 's4': 24};
    },
    meta: {multiEndpoint: true},
    exposes: [e.light_brightness().withLevelConfig().withEndpoint('l1'),
        exposes.numeric('ballast_minimum_level', ea.ALL).withValueMin(1).withValueMax(254)
            .withDescription('Specifies the minimum light output of the ballast')
            .withEndpoint('l1'),
        exposes.numeric('ballast_maximum_level', ea.ALL).withValueMin(1).withValueMax(254)
            .withDescription('Specifies the maximum light output of the ballast')
            .withEndpoint('l1'),
        exposes.enum('dimmer_mode', ea.ALL, ['RC', 'RL']).withDescription('Controls Capacitive or Inductive Dimming Mode')
            .withEndpoint('l1'),
        e.action(['on', 'off', 'brightness_move_up', 'brightness_move_down', 'brightness_stop', 'recall_*'])],
    configure: async (device, coordinatorEndpoint, logger) => {
        // Configure the dimmer actuator endpoint
        await extend.light_onoff_brightness().configure(device, coordinatorEndpoint, logger);
        const endpoint = device.getEndpoint(3);
        await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff', 'genLevelCtrl', 'lightingBallastCfg']);
        await reporting.onOff(endpoint);
        await reporting.brightness(endpoint);
        // Configure the four front switches
        device.endpoints.forEach(async (ep) => {
            if (21 <= ep.ID && ep.ID <= 22) {
                await reporting.bind(ep, coordinatorEndpoint, ['genOnOff', 'genLevelCtrl']);
            } else if (23 <= ep.ID && ep.ID <= 24) {
                await reporting.bind(ep, coordinatorEndpoint, ['genScenes']);
            }
        });
    },
    onEvent: async (type, data, device) => {
        // Record the factory default bindings for easy removal/change after deviceInterview
        if (type === 'deviceInterview') {
            const dimmer = device.getEndpoint(3);
            device.endpoints.forEach(async (ep) => {
                if (21 <= ep.ID && ep.ID <= 22) {
                    ep.addBinding('genOnOff', dimmer);
                    ep.addBinding('genLevelCtrl', dimmer);
                }
                if (23 <= ep.ID && ep.ID <= 24) {
                    ep.addBinding('genScenes', dimmer);
                }
            });
        }
    },
};

module.exports = definition;
```

I'm thinking I need to add `fz.occupancy` ([source](https://github.com/Koenkk/zigbee-herdsman-converters/blob/v15.0.88/converters/fromZigbee.js#L465)), but the multiple endpoints are making things a bit hard.

The clusters look like this, so I think I need to `Endpoint 37/Output clusters/msOccupancySensing` somehow, but I'm not sure where to start.

```

Endpoint 3
    Output clusters
        genOta
    Input clusters
        genBasic
        genIdentify
        genGroups
        genScenes
        genOnOff
        genLevelCtrl
        lightingBallastCfg
        haDiagnostic
Endpoint 37
    Output clusters
        genIdentify
        genOnOff
        genLevelCtrl
        msOccupancySensing
    Input clusters
        genBasic
        genIdentify
        msIlluminanceMeasurement
        msOccupancySensing
        haDiagnostic
        elkoOccupancySettingClusterServer
Endpoint 242
    Output clusters
        greenPower
```

**Comment by torroa** [comment]
Try this external converter.

It does not include luminance level or ocupancy sensor:

const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
zigbeeModel: ['NHMOTION/DIMMER/1'],
model: 'EKO06985',
vendor: 'Elko',
description: 'Elko plus - SmartPir dimmer',
fromZigbee: [fz.on_off, fz.brightness, fz.level_config, fz.schneider_lighting_ballast_configuration, fz.command_recall,
fz.command_on, fz.command_off, fz.command_move, fz.command_stop],
toZigbee: [tz.light_onoff_brightness, tz.level_config, tz.ballast_config, tz.schneider_dimmer_mode],
endpoint: (device) => {
return {'l1': 3, 's1': 21, 's2': 22, 's3': 23, 's4': 24};
},
meta: {multiEndpoint: true},
exposes: [e.light_brightness().withLevelConfig().withEndpoint('l1'),
exposes.numeric('ballast_minimum_level', ea.ALL).withValueMin(1).withValueMax(254)
.withDescription('Specifies the minimum light output of the ballast')
.withEndpoint('l1'),
exposes.numeric('ballast_maximum_level', ea.ALL).withValueMin(1).withValueMax(254)
.withDescription('Specifies the maximum light output of the ballast')
.withEndpoint('l1'),
exposes.enum('dimmer_mode', ea.ALL, ['RC', 'RL']).withDescription('Controls Capacitive or Inductive Dimming Mode')
.withEndpoint('l1'),
e.action(['on', 'off', 'brightness_move_up', 'brightness_move_down', 'brightness_stop', 'recall_*'])],
configure: async (device, coordinatorEndpoint, logger) => {
// Configure the dimmer actuator endpoint
await extend.light_onoff_brightness().configure(device, coordinatorEndpoint, logger);
const endpoint = device.getEndpoint(3);
await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff', 'genLevelCtrl', 'lightingBallastCfg']);
await reporting.onOff(endpoint);
await reporting.brightness(endpoint);
// Configure the four front switches
device.endpoints.forEach(async (ep) => {
if (21 <= ep.ID && ep.ID <= 22) {
await reporting.bind(ep, coordinatorEndpoint, ['genOnOff', 'genLevelCtrl']);
} else if (23 <= ep.ID && ep.ID <= 24) {
await reporting.bind(ep, coordinatorEndpoint, ['genScenes']);
}
});
},
onEvent: async (type, data, device) => {
// Record the factory default bindings for easy removal/change after deviceInterview
if (type === 'deviceInterview') {
const dimmer = device.getEndpoint(3);
device.endpoints.forEach(async (ep) => {
if (21 <= ep.ID && ep.ID <= 22) {
ep.addBinding('genOnOff', dimmer);
ep.addBinding('genLevelCtrl', dimmer);
}
if (23 <= ep.ID && ep.ID <= 24) {
ep.addBinding('genScenes', dimmer);
}
});
}
},
};

module.exports = definition;

**Comment by FSund** [comment]
I've given this a second attempt, and by combining the converter for `WDE002961` (push button dimmer) and `545D6306` (PIR with relay), I've come up with the following.

This supports on/off, dimming, ballast config, occupancy, and light level/illuminance.

```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const e = require('zigbee-herdsman-converters/lib/exposes').presets;
const ea = require('zigbee-herdsman-converters/lib/exposes').access;
const reporting = require('zigbee-herdsman-converters/lib/reporting');

const definition = {
    zigbeeModel: ['NHMOTION/UNIDIM/1'],
    model: 'EKO06984',
    vendor: 'ELKO',
    description: 'SmartPir with push dimmer',
    fromZigbee: [
        fz.on_off, 
        fz.brightness, 
        fz.illuminance,
        fz.level_config, 
        fz.schneider_lighting_ballast_configuration,
        fz.occupancy, 
    ],
    toZigbee: [
        tz.light_onoff_brightness, 
        tz.level_config, 
        tz.ballast_config, 
        tz.schneider_dimmer_mode,
    ],
    exposes: [
        e.light_brightness().withLevelConfig(),
        e.occupancy(),
        e.illuminance_lux(),
        e.illuminance(),
        e
            .numeric('ballast_minimum_level', ea.ALL)
            .withValueMin(1)
            .withValueMax(254)
            .withDescription('Specifies the minimum light output of the ballast'),
        e
            .numeric('ballast_maximum_level', ea.ALL)
            .withValueMin(1)
            .withValueMax(254)
            .withDescription('Specifies the maximum light output of the ballast'),
        e
            .enum('dimmer_mode', ea.ALL, ['RC', 'RL'])
            .withDescription('Controls capacitive or inductive dimming mode')
    ],
    configure: async (device, coordinatorEndpoint) => {
        const endpoint = device.getEndpoint(3);
        const binds = ['genOnOff', 'genLevelCtrl', 'lightingBallastCfg'];
        await reporting.bind(endpoint, coordinatorEndpoint, binds);
        await reporting.onOff(endpoint);
        await reporting.brightness(endpoint);

        const endpoint37 = device.getEndpoint(37);
        const binds37 = ['msIlluminanceMeasurement', 'msOccupancySensing'];
        await reporting.bind(endpoint37, coordinatorEndpoint, binds37);
        await reporting.occupancy(endpoint37);
        await reporting.illuminance(endpoint37);
    },
    whiteLabel: [
        {vendor: 'ELKO', model: 'EKO06985'},
        {vendor: 'ELKO', model: 'EKO06986'},
    ],
};

module.exports = definition;
```

I'll do some more testing to confirm everything works okay, and then submit a PR in a couple of days. Please let me know if you have any issues with it!

Draft PR here: https://github.com/Koenkk/zigbee-herdsman-converters/pull/8187

**Reply by torroa** [reply]
Have you tried to include fz & tz occupancy_timeout ?

**Reply by FSund** [reply]
Yes, sorry, I should have specified. I have tried that, but I get UNSUPPORTED_FEATURE (or perhaps `UNSUPPORTED_ATTRIBUTE`?) errors when I try to set occupancy timeout.
